[FrameworkBundle][Routing] Add a new tag to be able to use a private service as a service route loader

fancyweb opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/30402
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/11337

#eufossa

This PR adds a new tag `routing.route_loader` autoconfigured thanks to a new `ServiceRouterLoader` interface to be able to use private route loader services.

The deprecation layer is done through a temporary container that will be removed in 5.0.

TODO :
- [x] Changelog and upgrade entry
- [x] Doc PR

fancyweb committed:
[FrameworkBundle][Routing] Add a new tag to be able to use a private service as a service route loader

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderContainer.php:
```diff
@@ -0,0 +1,53 @@
+use Psr\Container\ContainerInterface;
+
+/**
+ * @internal
```
I'd suggest adding `@deprecated since Symfony 4.3, to be removed in 5.0`

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderContainer.php:
```diff
@@ -0,0 +1,53 @@
+            return $this->serviceLocator->get($id);
+        }
+
+        @trigger_error(sprintf('Not tagging the "%s" service with the "%s" tag is deprecated since Symfony 4.3.', $id, $this->serviceRouterLoaderTag), E_USER_DEPRECATED);
```
Registering a routing loader without tagging it... and will be required in 5.0.

fancyweb replied:
Registering the routing loader "%s" without tagging it with the "%s" tag is deprecated since Symfony 4.3 and will be required in Symfony 5.0. ?

Or the service id won't be visible anymore.

nicolas-grekas replied:
sure

nicolas-grekas reviewed (commented):
changelog+upgrade files need an update also

fancyweb commented:
@nicolas-grekas I added the UPGRADE entries already.

For the CHANGELOG of the Routing component there is nothing to say IMO. We added a marker interface, an instantly @deprecated / @internal class, and a compiler pass used by the Framework Bundle 😕

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/Tests/Kernel/MicroKernelTraitTest.php:
```diff
@@ -39,4 +42,21 @@ public function testAsEventSubscriber()
         $this->assertSame('It\'s dangerous to go alone. Take this ⚔', $response->getContent());
     }
+
+    public function testRoutingRouteLoaderTagIsAdded(): void
```
IMO we don’t use void return type, or?

cc @nicolas-grekas

nicolas-grekas replied:
on tests, sure! should be removed

fancyweb replied:
I don't mind removing it but for your information there are already 50 tests with the `void` return type in the code. Should we remove them?

OskarStark commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderContainer.php:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @internal
+ */
+class ServiceRouterLoaderContainer implements ContainerInterface
```
You should add an @author tag

fancyweb replied:
Is it mandatory?

nicolas-grekas replied:
it's not, up to you

OskarStark commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderInterface.php:
```diff
@@ -0,0 +1,16 @@
+
+namespace Symfony\Component\Routing\Loader\DependencyInjection;
+
+interface ServiceRouterLoaderInterface
```
You should add an @author tag

nicolas-grekas replied:
should add a docblock telling what this is about

nicolas-grekas commented on src/Symfony/Component/Routing/DependencyInjection/ServiceRouterLoaderPass.php:
```diff
@@ -0,0 +1,54 @@
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\Routing\Loader\DependencyInjection\ServiceRouterLoaderContainer;
+
+class ServiceRouterLoaderPass implements CompilerPassInterface
```
the class should be marked internal + have the same note as `ServiceRouterLoaderContainer`

fancyweb replied:
Ok for the internal but not for the note : this Compiler pass will not be removed in 5.0. It will just just be changed to directly inject the private service locator instead of our temporary container.

nicolas-grekas replied:
unless we can use the new `tagged_locator` argument type and get rid of this class :)

nicolas-grekas replied:
can't we right now btw?

fancyweb replied:
I did not know about this new feature. I will check it later this week and let you know.

fancyweb replied:
Sucessfully used a tagged locator, to be reviewed!

nicolas-grekas commented on src/Symfony/Component/Routing/Tests/Loader/DependencyInjection/ServiceRouterLoaderContainerTest.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp(): void
```
breaks with phpunit before 8 + php before 7.1: void should be removed I think

nicolas-grekas replied:
all of them :)

fancyweb replied:
I will remove it because there are no other instances of this in the code, but for my knowledge, how could that break with PHPUnit < 8 ? Because I actually always add strict return types even when parent / interface does not have it yet and I never had problems 😕 

For the PHP version I think it's fine since the minimum supported version on master is 7.1.3.

nicolas-grekas replied:
> add strict return types even when parent / interface does not have it yet and I never had problems

because you run this with PHP 7.2 - but using PHP 7.1, that would break, isn't it?

fancyweb replied:
It doesn't break (https://3v4l.org/QbQ6a) but it's not important.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -130,6 +131,7 @@ public function build(ContainerBuilder $container)
         $this->addCompilerPassIfExists($container, AddAutoMappingConfigurationPass::class);
         $container->addCompilerPass(new RegisterReverseContainerPass(true));
         $container->addCompilerPass(new RegisterReverseContainerPass(false), PassConfig::TYPE_AFTER_REMOVING);
+        $container->addCompilerPass(new ServiceRouterLoaderPass());
```
should use `addCompilerPassIfExists` as the class will be removed in 5.0

javiereguiluz commented:
Ping @fancyweb. Will you have some time to make the last requested changes? It'd be nice to have this for Symfony 4.3. Thanks!

fancyweb commented:
I will definitely finish this work before the end of the month.

fancyweb commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml:
```diff
@@ -40,9 +40,14 @@
         <service id="routing.loader.service" class="Symfony\Component\Routing\Loader\DependencyInjection\ServiceRouterLoader">
             <tag name="routing.loader" />
-            <argument type="service" id="service_container" />
+            <argument type="service" id="routing.loader.service.container" />
```
In 5.0, replace this argument with the tagged locator.

chalasr commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -423,6 +424,9 @@ public function load(array $configs, ContainerBuilder $container)
         }
+
+        $container->registerForAutoconfiguration(ServiceRouterLoaderInterface::class)
+            ->addTag('routing.route_loader');
```
This new tag needs to be added to `UnusedTagsPass::$whitelist`

chalasr commented:
Although I find `route loader` more clear, naming should be consistent between the interface and the tag (`RouterLoader` vs `RouteLoader`). We should stick with the current naming and rename the tag to `routing.router_loader` IMHO

fabpot commented on UPGRADE-4.3.md:
```diff
@@ -123,6 +123,7 @@ Routing
    options have been deprecated.
  * Implementing `Serializable` for `Route` and `CompiledRoute` is deprecated; if you serialize them, please
    ensure your unserialization logic can recover from a failure related to an updated serialization format
+ * Not tagging the route loader services with `routing.router_loader` has been deprecated.
```
`route loader` vs `router_loader`

fabpot replied:
I think router loader is way better.

fancyweb replied:
In this case, if the "public" name is now router loader, I guess we should update the whole documentation page (https://symfony.com/doc/current/routing/custom_route_loader.html)

fabpot replied:
and I meant `route loader`, sorry

fancyweb replied:
Do you mean you want to move back to the original tag name `routing.route_loader`?

chalasr replied:
Works for me, let's rename the interface + existing implementation, I suggest deprecating `ServiceRouterLoader` in favor of `ContainerRouteLoader`.

fancyweb replied:
WDYT of renaming everything in another PR after this one is merged for clarity ? First, finish this one (I just need to rename the tag back to `routing.route_loader`). Then, another one of pure refacto.

fancyweb replied:
friendly ping @fabpot @chalasr, would be cool to have this in 4.3

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderContainer.php:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @internal
+ *
+ * @deprecated since Symfony 4.3, to be removed in 5.0
```
`@internal, to be removed in 5.0`

fancyweb replied:
👌 but https://github.com/symfony/symfony/pull/30926#discussion_r272832533 ^^

nicolas-grekas replied:
oups :) then let's add both!

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php:
```diff
@@ -69,14 +69,23 @@ public function registerContainerConfiguration(LoaderInterface $loader)
+            $kernelDefinition = $container->getDefinition('kernel');
+
+            $kernelDefinition->addTag('routing.router_loader', [
+                'service_id' => 'kernel',
```
Do we have existing terminology for an attribute referencing a service?
What about using "id"?

fancyweb replied:
Tagged locators being new, we don't currently have usages in the core. We can use "id" here for sure, but the names mights vary depending on the tags and what they do I guess. Multiples services could be referenced for example.

nicolas-grekas replied:
Tagged locators are a new way to reference tags, but they don't inherently allow building anything new.
Everything the provide is and was possible using compiler passes.
The practice of using attributes to reference ids is not new AFAIK - or is nothing surprising even with 2.0 :)
BUT this made me realize #31463 would be a better behavior that this PR could the leverage.

fancyweb replied:
I actually made the mistake to look only for tagged locators, not in compiler passes. Thank you for your explanations.  https://github.com/symfony/symfony/pull/31463 is a great idea! I will update the code here accordingly.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml:
```diff
@@ -40,9 +40,14 @@
 
+        <service id="routing.loader.service.container" class="Symfony\Component\Routing\Loader\DependencyInjection\ServiceRouterLoaderContainer">
+            <argument type="service" id="service_container" />
+            <argument type="tagged_locator" tag="routing.router_loader" index-by="service_id" />
```
Actually, why do we need an index-by attribute at all?
Wouldn't adding a compiler pass allow us to remove the attribute altogether?
I may miss something :)

fancyweb replied:
The `index-by` attribute is needed to stay backward compatible with the `MicroKernelTrait` that uses an id for the kernel service in its router loader resource configuration (`kernel::loadRoutes`). That also allow everyone to be able to use service ids and not FQCN which is good IMO. I think we might use `static::class.'::loadRoutes'` here but is it backward compatible?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -428,6 +429,9 @@ public function load(array $configs, ContainerBuilder $container)
         }
+
+        $container->registerForAutoconfiguration(ServiceRouterLoaderInterface::class)
+            ->addTag('routing.router_loader');
```
but adding the tag doesn't work, because the tag needs an attribute right now, does it?
that's why removing the attribute on the tag might be better also?

fancyweb replied:
The tag works because the attribute `service_id` is a fallback. The default behavior is to use the FQCN.

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderInterface.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Marker interface for router loader services.
+ */
+interface ServiceRouterLoaderInterface
```
why the `Service` prefix?
wouldn't `RouterLoaderInterface` be a better name?

chalasr replied:
`RouteLoaderInterface` according to https://github.com/symfony/symfony/pull/30926#discussion_r281145005

fancyweb replied:
The `Service` prefix was to stay consistent with the existing `ServiceRouterLoader` class.

nicolas-grekas replied:
ok, so this confirms the prefix should be removed: "service" is an implementation detail that the interface shouldn't account for. And "Container" is the best practice now.
I agree with @chalasr's comments.

fancyweb replied:
Yes I have talked with @chalasr and I will deprecate / rename everyting directly in this PR 😄

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoaderContainer.php:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @deprecated since Symfony 4.3, to be removed in 5.0
+ */
+class ServiceRouterLoaderContainer implements ContainerInterface
```
`RouterLoaderContainer`?

chalasr replied:
I suggested `ContainerRouteLoader`, consistent with e.g. https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/CommandLoader/ContainerCommandLoader.php

nicolas-grekas commented:
#31463 is now merged, this can be rebased to leverage it and take comments into account. Thanks :)

fancyweb commented:
I'm closing here becaus it's a mess and I will reopen new splitted PRs 👀
